src/Domain/Services/CreateAccount.php

```diff
@@ -0,0 +1,257 @@
+
+		$customer_id = $this->create_customer_account(
+			$order->get_billing_email(),
+			$order->get_billing_first_name(),
+			$order->get_billing_last_name()
+		);
+
+		// Log the customer in and associate with the order.
+		wc_set_customer_auth_cookie( $customer_id );
+		$order->set_customer_id( get_current_user_id() );
+	}
+
+	/**
+	 * Check request options and store (shop) config to determine if a user account
+	 * should be created as part of order processing.
+	 *
+	 * @param \WP_REST_Request $request The current request object being handled.
+	 *
+	 * @return boolean True if a new user account should be created.
+	 */
+	protected function should_create_customer_account( \WP_REST_Request $request ) {
+		if ( is_user_logged_in() ) {
+			// User is already logged in - no need to create an account.
+			return false;
+		}
+
+		// From here we know that the shopper is not logged in.
+
+		if ( false === filter_var( get_option( 'woocommerce_enable_guest_checkout' ), FILTER_VALIDATE_BOOLEAN ) ) {
+			// Store requires an account for all checkouts (purchases).
+			// Create an account independent of shopper option in $request.
+			// Note - checkbox is not displayed to shopper in this case.
+			return true;
+		}
+
+		// From here we know that the store allows guest checkout;
+		// shopper can choose whether they sign up (`should_create_account`).
+
+		if ( true === filter_var( $request['should_create_account'], FILTER_VALIDATE_BOOLEAN ) ) {
+			// User has requested an account as part of checkout processing.
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Convert an account creation error to an exception.
+	 *
+	 * @param \WP_Error $error An error object.
+	 *
+	 * @return Exception.
+	 */
+	private function map_create_account_error( \WP_Error $error ) {
+		switch ( $error->get_error_code() ) {
+			// WordPress core error codes.
+			case 'empty_username':
+			case 'invalid_username':
+			case 'empty_email':
+			case 'invalid_email':
+			case 'email_exists':
+			case 'registerfail':
+				return new \Exception( 'woocommerce_rest_checkout_create_account_failure' );
+		}
+
+		return new \Exception( 'woocommerce_rest_checkout_create_account_failure' );
+	}
+
+	/**
+	 * Create a new account for a customer (using a new blocks-specific PHP API).
+	 *
+	 * The account is created with a generated username. The customer is sent
+	 * an email notifying them about the account and containing a link to set
+	 * their (initial) password.
+	 *
+	 * Intended as a replacement for wc_create_new_customer in WC core.
+	 *
+	 * @throws \Exception If an error is encountered when creating the user account.
+	 *
+	 * @param string $user_email The email address to use for the new account.
+	 * @param string $first_name The first name to use for the new account.
+	 * @param string $last_name  The last name to use for the new account.
+	 *
+	 * @return int User id if successful
+	 */
+	public function create_customer_account( $user_email, $first_name, $last_name ) {
```

Comment:
I made this public purely so I could test it .. which is not ideal, would rather keep it protected/private. 

Keen for suggestions on how to wrangle this.